src/daft-writers/src/file.rs

```diff
@@ -0,0 +1,139 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use daft_micropartition::MicroPartition;
+use daft_table::Table;
+
+use crate::{FileWriter, WriterFactory};
+
+// TargetFileSizeWriter is a writer that writes in files of a target size.
+// It rotates the writer when the current file reaches the target size.
+struct TargetFileSizeWriter {
+    current_file_rows: usize,
+    current_writer: Box<dyn FileWriter<Input = Arc<MicroPartition>, Result = Option<Table>>>,
+    writer_factory: Arc<dyn WriterFactory<Input = Arc<MicroPartition>, Result = Option<Table>>>,
+    target_in_memory_file_rows: usize,
+    results: Vec<Table>,
+    partition_values: Option<Table>,
+}
+
+impl TargetFileSizeWriter {
+    fn new(
+        target_in_memory_file_rows: usize,
+        writer_factory: Arc<dyn WriterFactory<Input = Arc<MicroPartition>, Result = Option<Table>>>,
+        partition_values: Option<Table>,
+    ) -> DaftResult<Self> {
+        let writer: Box<dyn FileWriter<Input = Arc<MicroPartition>, Result = Option<Table>>> =
+            writer_factory.create_writer(0, partition_values.as_ref())?;
+        Ok(Self {
+            current_file_rows: 0,
+            current_writer: writer,
+            writer_factory,
+            target_in_memory_file_rows,
+            results: vec![],
+            partition_values,
+        })
+    }
+
+    fn rotate_writer(&mut self) -> DaftResult<()> {
+        if let Some(result) = self.current_writer.close()? {
+            self.results.push(result);
+        }
+        self.current_file_rows = 0;
+        self.current_writer = self
+            .writer_factory
+            .create_writer(self.results.len(), self.partition_values.as_ref())?;
+        Ok(())
+    }
+}
+
+impl FileWriter for TargetFileSizeWriter {
+    type Input = Arc<MicroPartition>;
+    type Result = Vec<Table>;
+
+    fn write(&mut self, input: &Arc<MicroPartition>) -> DaftResult<()> {
+        use std::cmp::Ordering;
+        match (input.len() + self.current_file_rows).cmp(&self.target_in_memory_file_rows) {
+            Ordering::Equal => {
+                self.current_writer.write(input)?;
+                self.rotate_writer()?;
+            }
+            Ordering::Greater => {
+                // Finish up the current writer first
+                let remaining_rows = self.target_in_memory_file_rows - self.current_file_rows;
+                let (to_write, mut remaining) = input.split_at(remaining_rows)?;
+                self.current_writer.write(&to_write.into())?;
+                self.rotate_writer()?;
+
+                // Write as many full files as possible
```

Comment:
i think it would be cleaner to put the `match` under a `loop` and have an `local_offset`.

Then for `Ordering::Equal` and `Ordering::Less` you can early return.